File: 12 December/12-15/speedCheck.js
function speedCheck(speedMph, speedLimitKph) {
  const warningOverspeed = 5;
  const milesToKmConversionRate = 1.60934;
  const speedKph = speedMph * milesToKmConversionRate;

  return speedKph <= speedLimitKph ? "Not Speeding" 
    : speedKph <= speedLimitKph + warningOverspeed ? "Warning" 
    : "Ticket";
}

// --- TEST SUITE ---

const testsText = `
1. speedCheck(30, 70) should return "Not Speeding".
2. speedCheck(40, 60) should return "Warning".
3. speedCheck(40, 65) should return "Not Speeding".
4. speedCheck(60, 90) should return "Ticket".
5. speedCheck(65, 100) should return "Warning".
6. speedCheck(88, 40) should return "Ticket".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);

File: 12 December/12-15/speed_check.py
def speed_check(speed_mph, speed_limit_kph):
    warning_overspeed = 5;
    miles_to_km_conversion_rate = 1.60934
    speed_kph = speed_mph * miles_to_km_conversion_rate

    return "Not Speeding" if speed_kph <= speed_limit_kph \
        else "Warning" if speed_kph <= speed_limit_kph + warning_overspeed \
        else "Ticket"

# --- TEST SUITE ---

tests_text = '''
1. speed_check(30, 70) should return "Not Speeding".
2. speed_check(40, 60) should return "Warning".
3. speed_check(40, 65) should return "Not Speeding".
4. speed_check(60, 90) should return "Ticket".
5. speed_check(65, 100) should return "Warning".
6. speed_check(88, 40) should return "Ticket".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)